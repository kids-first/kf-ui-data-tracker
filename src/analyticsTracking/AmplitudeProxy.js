import {Amplitude} from '@amplitude/react-amplitude';
import {memoize} from '@amplitude/react-amplitude/src/lib/memoize';
import debounce from 'lodash.debounce';
import {buttonTracking} from './eventUtils';
import {EVENT_CONSTANTS} from '../analyticsTracking';

/**
 *
 */
class AmplitudeProxy extends Amplitude {
  logToConsole = false;

  constructor(props) {
    super(props);
    if (typeof props.debounceInterval === 'number') {
      this.logEvent = debounce(this.dispatch, props.debounceInterval);
    } else {
      this.logEvent = this.dispatch;
    }

    this._renderPropParams = {
      getInstance: this.getAmplitudeInstance,
      logEvent: this.logEvent,
      instrument: this.instrument,
      EVENT_CONSTANTS,
      buttonTracking: buttonTracking(this.logEvent),
    };
  }

  // proxy our logging calls so we can hook
  // other analytics services into it
  dispatch = (eventType, eventProps, cb) => {
    if (this.logToConsole || this.props.logToConsole) {
      console.log(
        `AmplitudeProxy::dispatch eventType:${eventType}`,
        eventProps,
      );
    }

    const combinedEventProps = {
      ...this.getAmplitudeEventProperties(),
      ...(eventProps || {}),
    };

    if (
      process.env.NODE_ENV === 'test' ||
      process.env.NODE_ENV === 'CI' ||
      this.props.storeSessionEvents
    ) {
      /** log all events to sessionStorage */
      sessionStorage.setItem(
        'session_events',
        JSON.stringify([
          ...(JSON.parse(sessionStorage.getItem('session_events')) || []),
          {
            eventType,

            eventProps: combinedEventProps,
            utc_time: Date.now(),
          },
        ]),
      );
    }

    return this._makeLogEvent()(eventType, combinedEventProps, cb);
  };

  instrument = memoize((eventType, func, props = {}) => {
    return (...params) => {
      const retVal = typeof func === 'function' ? func(...params) : undefined;
      this.dispatch(eventType, props);

      return retVal;
    };
  });
}

export default AmplitudeProxy;
