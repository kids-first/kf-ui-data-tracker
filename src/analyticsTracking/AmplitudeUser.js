import amplitude from 'amplitude-js';
import jwtDecode from 'jwt-decode';

/**
 * convenience class to log user properties on Auth0 login success
 * @see https://help.amplitude.com/hc/en-us/articles/115002380567-User-Properties-Event-Properties#applying-user-properties-to-events
 */
class AmplitudeUser {
  instance;
  userId;
  auth_sub_arr;

  // decoded jwt values
  auth_user;

  // values to pluck from jwt in the format of [[jwt_prop_name, renmae_to], ... ]
  user_props = [
    ['https://kidsfirstdrc.org/roles', 'roles'],
    ['https://kidsfirstdrc.org/groups', 'studies'],
    ['https://kidsfirstdrc.org/permissions', 'permissions'],
  ];

  constructor(idToken, api_key) {
    amplitude.getInstance().init(api_key);
    this.instance = amplitude.getInstance();
    this.auth_user = jwtDecode(idToken);

    this.auth_sub_arr = this.auth_user.sub.split('|');

    this.setId();
    this.setUserProperies();
    return this;
  }

  /**
   * UUID for amplitude using the parsed value of the
   * auth provider sub field
   * @param {string} id
   */
  setId(id) {
    this.userId = this.auth_sub_arr[1];
    this.instance.setUserId(this.userId);
  }

  setUserProperies() {
    var identify = new this.instance.Identify();

    // pluck properties from dceoded jwt and rename
    this.user_props.forEach(prop => {
      identify.set(prop[1] || prop[0], this.auth_user[prop[0]]);
    });
    // send to amplitude
    this.instance.identify(identify);
  }
}

export default AmplitudeUser;
