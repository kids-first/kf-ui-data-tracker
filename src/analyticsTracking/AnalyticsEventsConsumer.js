import React from 'react';
import {Amplitude} from '@amplitude/react-amplitude';
import useAnalyticsTracking from './useAnalyticsTracking';

const AnalyticsEventsConsumer = ({children, eventProps}) => {
  const analyticsConstants = useAnalyticsTracking();

  return (
    <Amplitude
      eventProperties={inheritedProps => ({
        ...inheritedProps,
        ...eventProps,
      })}
    >
      {props => {
        const buttonTrackingProps = buttonName => ({
          onClick: e =>
            props.logEvent(analyticsConstants.MOUSE.CLICK, {
              button_name: buttonName,
            }),
          onMouseEnter: e =>
            props.logEvent(analyticsConstants.MOUSE.HOVER, {
              button_name: buttonName,
            }),
        });
        if (typeof children === 'function') {
          return children({...props, buttonTrackingProps, analyticsConstants});
        } else {
          return children || null;
        }
      }}
    </Amplitude>
  );
};

export default AnalyticsEventsConsumer;
