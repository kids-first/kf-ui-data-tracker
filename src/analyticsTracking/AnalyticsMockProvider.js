import React from 'react';
import AnalyticsTrackingProvider from './AnalyticsTrackingProvider';

import {Amplitude, LogOnMount} from '@amplitude/react-amplitude';

const AnaltyticsMockProvider = ({children}) => (
  <AnalyticsTrackingProvider>
    <Amplitude eventProperties={{scope: ['TEST']}}>
      <LogOnMount eventType={'TEST_MOUNT'} />
      {children}
    </Amplitude>
  </AnalyticsTrackingProvider>
);

export default AnaltyticsMockProvider;
