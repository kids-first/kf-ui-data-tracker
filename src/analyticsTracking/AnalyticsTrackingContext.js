import React from 'react';
import analyticsTrackingConstants from '../common/analyticsTrackingConstants';

const AnalyticsTrackingContext = React.createContext(
  analyticsTrackingConstants,
);
export default AnalyticsTrackingContext;
