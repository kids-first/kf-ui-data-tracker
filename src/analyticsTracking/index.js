export {default as useAnalyticsTracking} from './useAnalyticsTracking';
export {default as withAnalyticsTracking} from './withAnalyticsTracking';

export {
  default as AnalyticsTrackingProvider,
} from './AnalyticsTrackingProvider';

export {default as AmplitudeProxy} from './AmplitudeProxy';
export {default as EVENT_CONSTANTS} from '../common/analyticsTrackingConstants';
