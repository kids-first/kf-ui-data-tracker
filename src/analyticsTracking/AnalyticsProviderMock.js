import React from 'react';
import AnalyticsTrackingProvider from './AnalyticsTrackingProvider';
import AmplitudeProxy from './services/Amplitude/AmplitudeProxy';

const AnaltyticsMockProvider = ({children}) => (
  <AnalyticsTrackingProvider>
    <AmplitudeProxy eventProperties={{scope: ['TEST']}}>
      {children}
    </AmplitudeProxy>
  </AnalyticsTrackingProvider>
);

export default AnaltyticsMockProvider;
