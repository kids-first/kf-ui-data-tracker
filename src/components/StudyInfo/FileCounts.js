import React from 'react';
import PropTypes from 'prop-types';
import {Link} from 'react-router-dom';
import {List, Icon, Popup, Label} from 'semantic-ui-react';
import {versionState} from '../../common/enums';
import {withAnalyticsTracking} from '../../analyticsTracking';

/**
 * Displays file counts with total number and breaking down by each status
 * When no files exist, show buttons guiding user to upload files
 */
const FileCounts = ({
  files,
  title,
  history,
  hideIcon,
  tracking: {buttonTracking, popupTracking},
}) => {
  const states = files.map(
    ({node: {versions}}) => versions.edges[0].node.state,
  );
  const stateCounts = states.reduce((count, state) => {
    count[state] = (count[state] || 0) + 1;
    return count;
  }, {});

  return (
    <List horizontal>
      <List.Item
        as={Link}
        to={`/study/${title}/documents`}
        onClick={
          popupTracking({
            name: 'Files',
            content: `${files.length > 0 ? files.length : 'No'} files`,
            link: `/study/${title}/documents`,
          }).onClick
        }
        className={hideIcon && files.length === 0 ? 'text-red' : null}
        {...buttonTracking({
          button_text: `${files.length > 0 ? files.length : 'No'} files`,
          button_type: 'text link',
        })}
      >
        {!hideIcon && (
          <Icon
            name="file"
            color={files && files.length > 0 ? 'grey' : 'red'}
          />
        )}
        {files.length > 0 ? files.length : 'No'} files
      </List.Item>
      {Object.keys(versionState)
        .slice(0, 4)
        .map(
          state =>
            state in stateCounts && (
              <Popup
                inverted
                position="top center"
                size="small"
                content={versionState[state].title}
                key={state}
                trigger={
                  <List.Item
                    {...popupTracking({
                      name: versionState[state].title,
                      content: stateCounts[state],
                    })}
                  >
                    <Label
                      circular
                      empty
                      size="mini"
                      color={versionState[state].labelColor}
                    />{' '}
                    {stateCounts[state]}
                  </List.Item>
                }
              />
            ),
        )}
    </List>
  );
};

FileCounts.propTypes = {
  /** Array of file object*/
  files: PropTypes.array.isRequired,
  /** Study id used by forming the redirect url*/
  title: PropTypes.string.isRequired,
  /** Rect-router history object  */
  history: PropTypes.object,
};

FileCounts.defaultProps = {
  files: [],
};

export default withAnalyticsTracking(FileCounts);
