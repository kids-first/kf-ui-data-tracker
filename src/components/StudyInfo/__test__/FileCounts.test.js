import React from 'react';
import {render, cleanup} from 'react-testing-library';
import {MemoryRouter} from 'react-router-dom';
import {MockedProvider} from 'react-apollo/test-utils';
import {AnalyticsProviderMock} from '../../../analyticsTracking';
import FileCounts from '../FileCounts';
import studyByKfId from '../../../../__mocks__/kf-api-study-creator/responses/studyByKfId';

afterEach(cleanup);

it('renders FileCounts in default stage', () => {
  const files = studyByKfId.data.studyByKfId.files.edges;
  const studyId = studyByKfId.data.studyByKfId.kfId;
  const tree = render(
    <AnalyticsProviderMock>
      <MemoryRouter>
        <MockedProvider>
          <FileCounts files={files} title={studyId} />
        </MockedProvider>
      </MemoryRouter>
    </AnalyticsProviderMock>,
  );
  expect(tree.container).toMatchSnapshot();
});

it('renders FileFileCounts with empty file list', () => {
  const studyId = studyByKfId.data.studyByKfId.kfId;
  const tree = render(
    <AnalyticsProviderMock>
      <MemoryRouter>
        <MockedProvider>
          <FileCounts files={[]} title={studyId} />
        </MockedProvider>
      </MemoryRouter>
    </AnalyticsProviderMock>,
  );
  expect(tree.container).toMatchSnapshot();
});
