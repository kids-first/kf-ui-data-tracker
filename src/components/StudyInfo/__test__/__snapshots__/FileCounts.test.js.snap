// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FileCounts in default stage 1`] = `
<div>
  <a
    href="/study/SD_8WX8QQ06/documents"
  >
    <i
      aria-hidden="true"
      class="grey file icon"
    />
    2
     documents
  </a>
  <div
    class="ui horizontal list ml-15"
    role="list"
  >
    <div
      class="item"
      role="listitem"
    >
      <div
        class="ui orange mini circular empty label"
      />
       
      2
    </div>
  </div>
</div>
`;

exports[`renders FileFileCounts with empty file list 1`] = `
<div>
  <a
    href="/study/SD_8WX8QQ06/documents"
  >
    <i
      aria-hidden="true"
      class="grey file icon"
    />
    No
     documents
  </a>
  <div
    class="ui horizontal list ml-15"
    role="list"
  />
</div>
`;
